lib/elastic_apm/normalizers/rails/active_record.rb

```diff
@@ -57,10 +57,17 @@ def normalize(_transaction, _name, payload)
         private
 
         def subtype_for(payload)
-          cached_adapter_name(
-            payload[:connection]&.adapter_name ||
-              ::ActiveRecord::Base.connection_config[:adapter]
-          )
+          return cached_adapter_name(payload[:connection].adapter_name) if payload[:connection]
+
+          if payload[:connection_id]
+            begin
+              loaded_object = ObjectSpace._id2ref(payload[:connection_id])
+              return cached_adapter_name(loaded_object.adapter_name) if loaded_object.respond_to?(:adapter_name)
+            rescue RangeError # if connection object has somehow been garbage collected
```

Comment:
We return `unknown` if we run into any trouble getting the connection object or its adapter name. So I think we should also return `unknown` in the case we can't get the connection object because it has been garbage collected.

Reply:
I think you also removed a test for this code path. Was that intentional?
https://github.com/elastic/apm-agent-ruby/pull/953/commits/c776010ff157e203cf5c128055366e9f8c43c525#diff-54b1044cee5cfa350d537561412b450f0858a900078239714e4f503306491251L108

Reply:
> We return `unknown` if we run into any trouble getting the connection object or its adapter name. So I think we should also return `unknown` in the case we can't get the connection object because it has been garbage collected.

Unless I misunderstand the code path you're referring to, I believe this does what you want. the final non-early-return case in `#subtype_for` calls `cached_adapter_name(nil)` which will return the adapter name in the `UNKNOWN` constant. The only cases in which we won't reach this final return value are:

- If we have a `payload[:connection]` value and we grab it's adapter name
- If we have a `payload[:connection_id]` value, it loads an object correctly, _and_ that object responds to `:adapter_name`

Otherwise that final code path should be reached and we'll get `unknown` as the adapter name.

> I think you also removed a test for this code path. Was that intentional?
> [c776010#diff-54b1044cee5cfa350d537561412b450f0858a900078239714e4f503306491251L108](https://github.com/elastic/apm-agent-ruby/commit/c776010ff157e203cf5c128055366e9f8c43c525#diff-54b1044cee5cfa350d537561412b450f0858a900078239714e4f503306491251L108)

This one is sort of an oversight on my part. I'm still going to "remove" this test but I'll be replacing it with one or two that properly test the fallback to the `unknown` adapter.

Reply:
Ok I pushed up the new tests. I don't see a way to effectively test the code path where the object_id lookup raises the `RangeError` due to the object having been garbage collected.

We'd somehow have to manually garbage collect an object, which, even if I knew how to do, I don't think would be a good idea. Do you have any suggestions, or are you ok with that code path remaining untested?